### ISSUE TITLE
Some notes

### ISSUE THREAD

**Issue by markvanderloo** [r-lib/vctrs#43]
Since [comments are welcome](https://twitter.com/hadleywickham/status/1026204037765627904), some thoughts.

I think that having a list type that forces all objects to be of the same class can be really useful. I do have questions about introducing a new set of atomic types.

R has a type system that can already be confusing, there is 

- `typeof`, determining the R internal type
- `mode` is a type designation that's closer to user experience (see below)
- `storage mode` giving something closer to physical storage type (e.g. `double` for a value of type `numeric`)
- `class`  and `inherits` for basic types and (`S3`, `S4`, `RC`,...) class extensions and best for users.

Some examples
```r
> types <- function(x) c(class=class(x)
  , typeof=typeof(x), mode=mode(x), storage.mode=storage.mode(x))
> dat <- list(int=1L,real=pi, complex=2+3i, string = character(), cat=factor(), binary=as.raw(0), fun=function(){})

> dat <- list(integer=1L,real=pi, complex=2+3i
  , string = character(), categorical=factor(), binary=as.raw(0)
  , 'function'=function(){})
> t(sapply(dat, types))
            class       typeof      mode        storage.mode
integer     "integer"   "integer"   "numeric"   "integer"   
real        "numeric"   "double"    "numeric"   "double"    
complex     "complex"   "complex"   "complex"   "complex"   
string      "character" "character" "character" "character" 
categorical "factor"    "integer"   "numeric"   "integer"   
binary      "raw"       "raw"       "raw"       "raw"       
function    "function"  "closure"   "function"  "function"  
```

So the question is if another `atomic' type designation just for the benefit of stricter coercion rules will help users. In my experience, it is already uncomfortable having to explain novice users the difference between data frames and tibbles. This is necessary because they will run into both very quickly once they start doing actual work with R. When you are a new user, this is another thing to put on your mental stack that has nothing to do with _getting things done_. This is even more true for vector types. For example, until now, I can explain R's coercion rules fairly easily, by letting people experiment with a few lines of code like:

```r
c(10, "hello")
```
Then, when I ask them to explain what happened and why, this usually clears things up pretty quickly.

With these new vector types new users would have to grok a second type system, especially if you automatically translate everything that is (implicitly) translated to a tibble. [By the way, beyond R's usual behavior, I think that type casts should be asked for by the user explicitly (_Explicit is better than implicit_)].

So in conclusion, I do not feel that the benefits of stricter coercion rules outweigh the burden of having to cope with an extra system for atomic types for end-users. I see there may be benefits for (package) developers. So it should be developer-facing rather than user-facing.

**Comment by lionel-** [comment]
What do you mean by "another atomic type designation"? For bare types we are using R's internal type nomenclature (integer, double, ...), otherwise the class names (factor, data.frame, ...). This is consistent with type annotations in printed tibbles or type suffixes in purrr.

One of the main ideas in vctrs is using length-zero vectors as "prototypes" that are cheap to pass around. This way the engine can reason about types (e.g. compute the least-upper-bound type) without any cost. This is why you can use empty vectors to refer to types, e.g. `.type = integer()`. Base R uses a variant of this idea with `vapply()`.

**Comment by markvanderloo** [comment]
FWIW: `tibble` printing `dbl` on top of a numeric column is IMO a bad example. Printing `double` or `dbl` in `tibble` objects is already confusing for new user who just learned what `numeric` means.
  
My main objection is that I think is a bad idea to have two separate systems for what I call atomic types and what you call bare types (basically the representation of integers, real numbers, strings) that is _user facing_. It seriously complicates reasoning about code unless you must explicitly convert somewhere. 

I understand there are advantages for developers. But for end-users working interactively I believe it is confusing. And I have [argued elsewhere](https://stat.ethz.ch/pipermail/r-package-devel/2018q3/002945.html) that there are different trade-offs for different uses of code.

**Comment by hadley** [comment]
To be a little more precise, I'd say that:

* `typeof()` gives the [base type](https://adv-r.hadley.nz/base-types.html). It's not clear to me that this forms a type _system_, as the results you get from combining two vectors varies from place-to-place (e.g. `c(x, y)` and `c(unlist(x, y))` can give different results

* `mode()` and `storage.mode()` exist primarily for compatibility with S. I think they can be ignored in this discussion.

* S3 and S4 are object oriented systems that don't in general say how different classes of objects should be combined. Even getting the correct result of `c()` requires quite some thought because it only dispatches on the class of the first argument.

This precision may seem unnecessary but I think it's really important to agree on terms because otherwise we end up talking past each other. This imprecision is the reason that tibbles print `dbl` instead of `num`: base R uses numeric to mean [three different things](https://adv-r.hadley.nz/base-types.html#numeric-type), so to me it seemed better to

Apart from those technical distinctions, I agree with pretty much everything you are saying:

* You are absolutely right that these distinctions are developer facing, not user facing. The goal of this work is to come up with a system that is simple and more consistent so that a user can take what they know about one function (e.g. `dplyr::bind_rows()` and apply it to other function e.g. `tidyr::gather()`)

* Implicit coercions are dangerous, which is why vctrs makes far fewer of them than base R, and the ones we do make have been carefully considered (mostly to ensure that existing R users don't become frustrated with the need to become extra precise)

* `c()` is ok when you're working with atomic vectors. I'll add a few examples to the README shortly to show some of the problems that occur when you start working with more complex objects.

**Comment by hadley** [comment]
Base R comparisons now at https://github.com/r-lib/vctrs#compared-to-base-r. Hopefully that will help you understand the base R behaviour that I worry about.

**Comment by hadley** [comment]
Ah, yes, the emphasis is on _system_, not on types. vctrs currently contributes one new type (`list_of`); it's contribution is mostly an algebraic system for dealing with types. It's useful to get this feedback as I'm so close to the material it's hard to see how others will interpret it. I've added some more motivation higher up on the readme to hopefully help with this.

vctrs does form a commutative monoid, but I didn't want to brag 😛 I've chosen to use `NULL` as the identity element as it's often used to represent a vector of unknown type (i.e. as default value for function argument). This also justifies the return value of `vec_c()`.

I have also thought about the ordering of factor levels. I definitely don't want to take exactly the same approach as base R because it orders factor levels alphabetically, and the order of the letters in the alphabet varies from locale to locale (and the having the same code return different results on different systems seems like a bad principle to me). I _think_ what I actually want is to be more precise about what I mean by commutativity - it only applies to the "container" type, not the elements of the type. The same issue comes up with data frames, where you definitely don't want to alphabetise the variables, and ideally factors and data frames would rely on the same justification for their behaviour. But this is hand-wavy at the moment, and I'll work to make it more clear.

**Comment by markvanderloo** [comment]
I would like to dicuss the monoid idea a little further from a formal point of view as these are really useful concepts in programming (btw you have probably seen this, but if you didn't: I can really recommend the material of Bartosz Milevski [youtube](https://www.youtube.com/watch?v=I8LbkfSSR58&list=PLbgaMIhjbmEnaH_LTkxLI7FMa2HsnawM_), [blog](https://bartoszmilewski.com/2014/10/28/category-theory-for-programmers-the-preface/))

Let's look at a single type, like `integer`. In programming a type is just a nonempty set of allowed values. For R's `integer` the allowed values are sequences containing `int` values. Since `NULL` is not such a sequence, it does not belong to the type `integer`. The role of the neutral element is played by the object returned by `integer()`. So the conclusion is that `vec_c` is a monoid on the type we call `integer` in this sense.

Now, how about `vec_c` for integers and reals? Well, we need to formally define union type, say `integer|double` whose set of allowed values are sequences of either `integer` or `double`. When you call `vec_c(x,y)` the result is always again `integer|double`, so we are closed under this 'multiplication'. Now the question is, what is the neutral element? It depends on the coercion rule: if `vec_c` returns a double when fed with an integer and a double, then clearly the neutral element _must_ be the object returned by `integer`:

```r
# we have
vec_c(pi, integer()) == pi
# and also
vec_c(1L, integer()) == 1L
```
The other candidate would be `double()`, but that's not a neutral element because:
```r
vec_c(1L, double()) != 1L
```

So I think that it may be possible to avoid `NULL` as a neutral element. My instinct is that this is to be preferred because it is more like the empty set than a type (you can define a function on an empty set but you can never call it). So from a formal side I would probably avoid `NULL`. But, R's behavior does not always follow formal rules so from a practical viewpoint a different choice may be preferable. I have to work out a few more examples to really be sure. Interesting stuff!

**Comment by lionel-** [comment]
The identity of `vec_c()` cannot affect the final type. `vec_c(identity, TRUE)` must be the same as `vec_c(TRUE)` and `vec_c(identity, 1)` must be the same as `vec_c(1)`. This is why it can't be an empty vector and `NULL` makes sense. This is also consistent with base R.

**Comment by hadley** [comment]
I now have `unknown()`, which I'm probably going to change to `anything()`. It represents a vector of known length and unknown type, for example when loading an empty column from a csv file.

**Comment by krlmlr** [comment]
1. We could also describe the commutativity as:

    ```r
    identical(vec_ptype_abbr(vec_c(x, y)), vec_ptype_abbr(vec_c(y, x)))
    ```

2. Strictly, `vec_type2` is a partially defined function because it is defined only for some pairs of types. We can think about the error condition as an "infinity" element in our type system (with `NULL` as identity) to turn it into a function "defined" for all pairs.

3. It may be easier to reason about combining types if the operator is fully commutative (and not only on the "container" type). There are ways to make this system-independent, e.g. by comparing strings in the C locale.

    - Factor levels: For `x <- factor(1:2, levels = letters[2:1])` and `y <- factor(1:3, levels = letters[4:2])`, both `vec_c(x, y)` and `vec_c(y, x)` should lead to the levels `c("b", "a", "d", "c")` .

    - Time zones: Use the time zone with the lowest offset (system-dependent?) or perhaps with the lowest ranking of the time zone string

    - Data frame columns: Similar to factor levels.

    This comes at a cost (implementation and run time), though.
